### ISSUE TITLE
Muon: support complex parameters

### ISSUE THREAD

**Issue by eirikfagerbakke** [google-deepmind/optax#1494]
The Muon optimizer currently does not support complex parameters.
A very simple and natural extension would be to make the Newton-Schulz iterator push the gradients to be semi-unitary for complex parameters, rather than the current semi-orthogonalization.
I.e. $U \Sigma V^* \rightarrow  U V^*$, rather than the current  $U \Sigma V^\top \rightarrow  U V^\top$.
The only update needed is to add a `.conj()` in `_newton_schulz_iterator.`
```
def _newton_schulz_iterator(x: jax.Array, coeffs: jax.Array) -> jax.Array:
  # Implements Newton-Schulz step f(X) = c_0 X + c_1 (XX^T)X + c_2 (XX^T)^2X,
  # with quintic form f(X) = c_0 X + (c_1 A + c_2 AA)X, where A = XX^T.
  # The NS step has the property f(X) = f(X^T)^T. That is, we can get equivalent
  # result by transposing input and output. In particular, we may transpose X
  # when rows > cols for efficiency.
  a = x @ x.T.conj()
  b = coeffs[1] * a + coeffs[2] * a @ a
  return coeffs[0] * x + b @ x
```

I also think one would have to update the dual norm scaling to reflect the complex inner product, but I am not entirely sure what this part refers to in the referenced article.
```
if adaptive:
  # Scale the orthogonalized updates by the dual norm of the original
  # updates. See https://arxiv.org/abs/2409.20325 for the derivation.
  updates = jax.tree.map(
      lambda x, y: jnp.sum(x.conj() * y) * y, mu_hat, updates
  )
```

This simple extension does not affect the use on real parameters, as the `.conj()` will be ignored under `jax.jit` afaik.

**Comment by vroulet** [comment]
Thanks for raising this!
Do you have
1. a use example?
2. tests for what you propose?

**Comment by eirikfagerbakke** [comment]
> Thanks for raising this! Do you have
> 
> 1. a use example?

Yes! This would be applicable for models with complex parameters, such as e.g. complex-valued neural networks or the Fourier Neural Operator. Other optimizers in optax, such as Adam, already handle complex parameters.

> 2. tests for what you propose?

The reason why we need to conjugate is that the standard complex inner product differs from the one for real numbers, so orthogonality translates to unitarity.

Following the exposition given in [Keller Jordan's blog](https://kellerjordan.github.io/posts/muon/#proving-that-ns-iteration-orthogonalizes-the-update):
Let $G=U S V^*$ be the gradients and its SVD decomposition, and let the update step be given by $\phi(G):=aG + b(GG^\*)G + c(GG^\*)^2 G$. We can rewrite this as

$$
\begin{aligned}
G'
&= aG + b(GG^\*)G + c(GG^\*)^2 G \\
&= (aI + b GG^\* + c(GG^\*)^2) G \\
&= (aI + bUS^2U^\* + cUS^4U^\*)  USV^\* \\
&= U(aS + bS^3 + cS^5)V^\*.\\
&= U \phi (S)V^*
\end{aligned}
$$

Similarly to the real case, we want $\phi^N(S) \to 1$, as the singular values for unitary matrices are also 1. The only change needed is therefore to change the tranpose to a conjugate transpose.

For a simple numerical example, we can apply a few steps of the Newton-Schulz algorithm to a random complex matrix:

```
import jax
import jax.numpy as jnp

def current_newton_schulz(x, num_iters=10):
    coeffs = (2,-1.5,0.5) 

    x = x/jnp.linalg.norm(x)

    transposed = False
    if x.shape[0] > x.shape[1]:
        x = x.T
        transposed = True

    def step(i, x):
        a = x @ x.T
        b = coeffs[1] * a + coeffs[2] * (a @ a)
        return coeffs[0] * x + b @ x
    x = jax.lax.fori_loop(0, num_iters, step, x)
    if transposed:
        x = x.T
    return x

def new_newton_schulz(x, num_iters=10):
    coeffs = (2,-1.5,0.5)

    x = x/jnp.linalg.norm(x)
    
    transposed = False
    if x.shape[0] > x.shape[1]:
        x = x.T
        transposed = True

    def step(i, x):
        a = x @ x.T.conj() # adding .conj() here is the only change!
        b = coeffs[1] * a + coeffs[2] * (a @ a)
        return coeffs[0] * x + b @ x
    x = jax.lax.fori_loop(0, num_iters, step, x)
    if transposed:
        x = x.T
    return x

# G is a random complex matrix, meant to represent a weight matrix
G = jax.random.normal(jax.random.key(0), (6,5)) + 1j*jax.random.normal(jax.random.key(1), (6,5))

# Let's see if updates are orthognal/unitary
G_current_ns = current_newton_schulz(G)
G_new_ns = new_newton_schulz(G)
I = jnp.eye(5)

print(jnp.allclose(G_current_ns.T @ G_current_ns, I, atol=1e-7)) #False, does not orthogonalize. The entries of `G_current_ns` are `nan`.
print(jnp.allclose(G_new_ns.T.conj() @ G_new_ns, I, atol=1e-7)) #True, becomes unitary
```

**Comment by vroulet** [comment]
That looks pretty good! 
So I think you have all ready for a PR: add the conj and add the associated test of the newton schulz iterations.
Would you be willing to do that PR? If yes ping me when it's ready, otherwise I'll try to take a look in a month.
Thank you for the neat proposal

**Comment by eirikfagerbakke** [comment]
I have made the PR now! #1506 Let me know what you think:) @vroulet
